### PR TITLE
feat: custom android buffer

### PR DESCRIPTION
**Pull request by b4tchkn**
*Replace this paragraph with a description of what this PR is changing or adding, and why. Consider including before/after screenshots.*

*List which issues are fixed by this PR. You must list at least one issue.*

*If you had to change anything in the [flutter/tests] repo, include a link to the migration guide as per the [breaking change policy].*

## Pre-launch Checklist

- [ ] I read the [Contributor Guide] and followed the process outlined there for submitting PRs.
- [ ] I read the [Tree Hygiene] wiki page, which explains my responsibilities.
- [ ] I read and followed the [relevant style guides] and ran [the auto-formatter]. (Unlike the flutter/flutter repo, the flutter/plugins repo does use `dart format`.)
- [ ] I signed the [CLA].
- [ ] The title of the PR starts with the name of the plugin surrounded by square brackets, e.g. `[shared_preferences]`
- [ ] I listed at least one issue that this PR fixes in the description above.
- [ ] I updated `pubspec.yaml` with an appropriate new version according to the [pub versioning philosophy], or this PR is [exempt from version changes].
- [ ] I updated `CHANGELOG.md` to add a description of the change, [following repository CHANGELOG style].
- [ ] I updated/added relevant documentation (doc comments with `///`).
- [ ] I added new tests to check the change I am making, or this PR is [test-exempt].
- [ ] All existing and new tests are passing.

If you need help, consider asking for advice on the #hackers-new channel on [Discord].

[Contributor Guide]: https://github.com/flutter/plugins/blob/main/CONTRIBUTING.md
[Tree Hygiene]: https://github.com/flutter/flutter/wiki/Tree-hygiene
[relevant style guides]: https://github.com/flutter/plugins/blob/main/CONTRIBUTING.md#style
[CLA]: https://cla.developers.google.com/
[flutter/tests]: https://github.com/flutter/tests
[breaking change policy]: https://github.com/flutter/flutter/wiki/Tree-hygiene#handling-breaking-changes
[Discord]: https://github.com/flutter/flutter/wiki/Chat
[pub versioning philosophy]: https://dart.dev/tools/pub/versioning
[exempt from version changes]: https://github.com/flutter/flutter/wiki/Contributing-to-Plugins-and-Packages#version-and-changelog-updates
[following repository CHANGELOG style]: https://github.com/flutter/flutter/wiki/Contributing-to-Plugins-and-Packages#changelog-style
[the auto-formatter]: https://github.com/flutter/plugins/blob/main/script/tool/README.md#format-code
[test-exempt]: https://github.com/flutter/flutter/wiki/Tree-hygiene#tests

### CODE REVIEW

**Review by ostk0069 — APPROVED** [review]

File: packages/video_player/video_player_android/android/src/main/java/io/flutter/plugins/videoplayer/VideoPlayer.java [hunk 1]
```diff
@@ -72,7 +81,16 @@ final class VideoPlayer {
     this.textureEntry = textureEntry;
     this.options = options;
 
-    ExoPlayer exoPlayer = new ExoPlayer.Builder(context).build();
+    DefaultLoadControl.Builder defaultLoadControl = new DefaultLoadControl.Builder();
+    defaultLoadControl.setBufferDurationsMs(
```
Comment (b4tchkn) [inline packages/video_player/video_player_android/android/src/main/java/io/flutter/plugins/videoplayer/VideoPlayer.java:85]: もともとはこれしてなかったので、中でデフォルトの値が使われていました。
デフォルトの値たち
https://github.com/google/ExoPlayer/blob/a9444c880230d2c2c79097e89259ce0b9f80b87d/library/core/src/main/java/com/google/android/exoplayer2/DefaultLoadControl.java#L35-L56

File: packages/video_player/video_player_android/android/src/main/java/io/flutter/plugins/videoplayer/VideoPlayerPlugin.java [hunk 2]
```diff
@@ -223,6 +228,25 @@ public void setMixWithOthers(MixWithOthersMessage arg) {
     options.mixWithOthers = arg.getMixWithOthers();
   }
 
+  @Override
+  public void setBuffer(BufferMessage arg) {
+    if (arg == null) return;
+    VideoPlayerBuffer buffer = new VideoPlayerBuffer();
+    buffer.minBufferMs = (arg.getMinBufferMs() == null)
```
Comment (b4tchkn) [inline packages/video_player/video_player_android/android/src/main/java/io/flutter/plugins/videoplayer/VideoPlayerPlugin.java:235]: nullのときはもともとのデフォルトの値を使うようにしました。
toIntExactはintに変換できない32ビット以上のlongがくるとExceptionを出しますが、運用上そんな大きな値は入れないので大丈夫かなと思っています。

File: packages/video_player/video_player_avfoundation/lib/src/avfoundation_video_player.dart [hunk 3]
```diff
@@ -175,6 +175,13 @@ class AVFoundationVideoPlayer extends VideoPlayerPlatform {
         .setMixWithOthers(MixWithOthersMessage(mixWithOthers: mixWithOthers));
   }
 
+  @override
+  Future<void> setBuffer(Buffer buffer) {
```
Comment (b4tchkn) [inline packages/video_player/video_player_avfoundation/lib/src/avfoundation_video_player.dart:179]: iOSではセット部分なにもしない実装です。

File: packages/video_player/video_player_android/lib/src/android_video_player.dart [hunk 4]
```diff
@@ -171,6 +171,19 @@ class AndroidVideoPlayer extends VideoPlayerPlatform {
         .setMixWithOthers(MixWithOthersMessage(mixWithOthers: mixWithOthers));
   }
 
+  @override
+  Future<void> setBuffer(Buffer buffer) {
+    return _api.setBuffer(
```
Comment (ostk0069) [inline packages/video_player/video_player_android/lib/src/android_video_player.dart:176]: 値が0だったらそもそもsetしないみたいなロジックはない認識？
defaultの値を使いたい時はremote configの値をdefaultにする感じ？

File: packages/video_player/video_player_platform_interface/lib/video_player_platform_interface.dart [hunk 5]
```diff
@@ -108,6 +108,62 @@ abstract class VideoPlayerPlatform extends PlatformInterface {
...
+  ///
+  /// The [bufferForPlaybackAfterRebufferMs] argument can be null.
+  Buffer({
+    this.minBufferMs,
+    this.maxBufferMs,
+    this.bufferForPlaybackMs,
+    this.bufferForPlaybackAfterRebufferMs,
+  });
+
+  /// This value is only used in Android.
```
Comment (b4tchkn) [inline packages/video_player/video_player_platform_interface/lib/video_player_platform_interface.dart:138]: このへんの英語のコメントはExoPlayerとAVPlayerからコピーしてきてます

File: packages/video_player/video_player_android/android/src/main/java/io/flutter/plugins/videoplayer/Messages.java [hunk 6]
```diff
@@ -478,34 +478,115 @@ public static final class Builder {
...
+      toMapResult.put("minBufferMs", minBufferMs);
+      toMapResult.put("maxBufferMs", maxBufferMs);
+      toMapResult.put("bufferForPlaybackMs", bufferForPlaybackMs);
+      toMapResult.put("bufferForPlaybackAfterRebufferMs", bufferForPlaybackAfterRebufferMs);
+      return toMapResult;
+    }
+    static @NonNull BufferMessage fromMap(@NonNull Map<String, Object> map) {
+      BufferMessage pigeonResult = new BufferMessage();
+      Object minBufferMs = map.get("minBufferMs");
+      pigeonResult.setMinBufferMs((minBufferMs == null) ? null : ((minBufferMs instanceof Integer) ? (Integer)minBufferMs : (Long)minBufferMs));
```
Comment (Yokoi-K) [inline packages/video_player/video_player_android/android/src/main/java/io/flutter/plugins/videoplayer/Messages.java:549]: どういう時にIntegerとLongが異なるのですか？

**Review by Yokoi-K — APPROVED** [review]
Java良さそうだと思いました

File: packages/video_player/video_player_android/android/src/main/java/io/flutter/plugins/videoplayer/Messages.java (hunk 6 above)
Comment (b4tchkn) [reply packages/video_player/video_player_android/android/src/main/java/io/flutter/plugins/videoplayer/Messages.java:549]: ここ自動生成されたとこなのでわからないですね、、、
pigeon側の挙動です。

File: packages/video_player/video_player_android/lib/src/android_video_player.dart (hunk 4 above)
Comment (b4tchkn) [reply packages/video_player/video_player_android/lib/src/android_video_player.dart:176]: そうだね。
AndroidもiOSもわかりやすいようにそうしようと思う。
なにも渡したくないときもある（過去動画とかTopの動画とか）と思うからnullableにしてnullだったらデフォルトの値使うようにしました。

File: packages/video_player/video_player_android/android/src/main/java/io/flutter/plugins/videoplayer/VideoPlayer.java [hunk 7]
```diff
@@ -72,7 +81,16 @@ final class VideoPlayer {
...
+        options.buffer.minBufferMs,
+        options.buffer.maxBufferMs,
+        options.buffer.bufferForPlaybackMs,
+        options.buffer.bufferForPlaybackAfterRebufferMs
+    );
```
Comment (akihisasengoku) [inline packages/video_player/video_player_android/android/src/main/java/io/flutter/plugins/videoplayer/VideoPlayer.java:90]:
```suggestion
    DefaultLoadControl defaultLoadControl = new DefaultLoadControl.Builder().setBufferDurationsMs(
        options.buffer.minBufferMs,
        options.buffer.maxBufferMs,
        options.buffer.bufferForPlaybackMs,
        options.buffer.bufferForPlaybackAfterRebufferMs
    ).build();
    defaultLoadControl.setBufferDurationsMs(
        options.buffer.minBufferMs,
        options.buffer.maxBufferMs,
        options.buffer.bufferForPlaybackMs,
        options.buffer.bufferForPlaybackAfterRebufferMs
    );
```

**Review by akihisasengoku — APPROVED** [review]

File: packages/video_player/video_player_android/android/src/main/java/io/flutter/plugins/videoplayer/VideoPlayer.java (hunk 7 above)
Comment (b4tchkn) [reply packages/video_player/video_player_android/android/src/main/java/io/flutter/plugins/videoplayer/VideoPlayer.java]: 今後も `build()` する前になにかsetしたくなった場合、今のほうが書きやすいと思うのでこのままでいきたいです 🙏

**Pull request merged**